dialogue-core/src/main/java/com/palantir/dialogue/core/PinUntilErrorChannel.java

```diff
@@ -0,0 +1,270 @@
+        this.nodeList = nodeList;
+    }
+
+    static PinUntilErrorChannel pinUntilError(List<LimitedChannel> channels) {
+        Preconditions.checkArgument(!channels.isEmpty(), "List of channels must not be empty");
+        ReshufflingNodeList shufflingNodeList =
+                new ReshufflingNodeList(channels, SafeThreadLocalRandom.get(), System::nanoTime);
+        return new PinUntilErrorChannel(shufflingNodeList);
+    }
+
+    /**
+     * This strategy results in unbalanced server utilization, especially after an upstream restarts.
+     * @deprecated prefer {@link #pinUntilError}
+     */
+    @Deprecated
+    static PinUntilErrorChannel pinUntilErrorWithoutReshuffle(List<LimitedChannel> channels) {
+        Preconditions.checkArgument(!channels.isEmpty(), "List of channels must not be empty");
+        ConstantNodeList constantNodeList = new ConstantNodeList(channels, SafeThreadLocalRandom.get());
+        return new PinUntilErrorChannel(constantNodeList);
+    }
+
+    @Override
+    public Optional<ListenableFuture<Response>> maybeExecute(Endpoint endpoint, Request request) {
+        int currentIndex = currentHost.get();
+        LimitedChannel channel = nodeList.get(currentIndex);
+
+        Optional<ListenableFuture<Response>> maybeFuture = channel.maybeExecute(endpoint, request);
+        if (!maybeFuture.isPresent()) {
+            OptionalInt next = markCurrentHostAsBad(currentIndex);
+            debugLogCurrentChannelRejected(currentIndex, channel, next);
+            return Optional.empty(); // if the caller retries immediately, we'll get the next host
+        }
+
+        ListenableFuture<Response> future = maybeFuture.get();
+        return Optional.of(DialogueFutures.addDirectCallback(future, new FutureCallback<Response>() {
+            @Override
+            public void onSuccess(Response response) {
+                if (response.code() >= 300) {
+                    OptionalInt next = markCurrentHostAsBad(currentIndex);
+                    debugLogReceivedErrorStatus(currentIndex, channel, response, next);
+                    // TODO(dfox): handle 308 See Other somehow, as we currently don't have a host -> channel mapping
+                }
+            }
+
+            @Override
+            public void onFailure(Throwable throwable) {
+                OptionalInt next = markCurrentHostAsBad(currentIndex);
+                debugLogReceivedThrowable(currentIndex, channel, throwable, next);
+            }
+        }));
+    }
+
+    /**
+     * If we have some reason to think the currentIndex is bad, we want to move to the next host. This is done with a
+     * compareAndSet to ensure that out of order responses which signal information about a previous host don't kick
+     * us off a good one.
+     */
+    private OptionalInt markCurrentHostAsBad(int currentIndex) {
+        int nextIndex = Math.max(currentIndex + 1, 0); // we want Integer.MAX_VALUE to wrap around to zero
```

Comment:
interestingly I think the round robin channel has this same bug, because if a list has 10 nodes, and we end up doing:

```
(Integer.MAX_VALUE + 1) % 10 = -8
```

And I'm _pretty_ sure arraylists don't like negative indexes.

Reply:
Can we write a test to expose it?